initial attempt at solidjs ext support

kristianmandrup opened the pull request:
Initial attempt at adding SolidJS Nx extension - has yet to be fully tested.

DominikPieper commented:
Good morning,
wow a new plugin is a really great contribution, thanks so far. The build actions are now able to run. :-)

kristianmandrup commented:
Awesome! Thanks :) You can have a look at https://how-to.dev/how-to-start-a-project-with-solidjs to get a "solid" impression. Has a bunch of articles including https://how-to.dev/how-to-build-solidjs-application-with-esbuild which would be a logical next step after initial basic supports completely works. SolidJS leverages vite.

I'm not sure if I need to add the scripts to `package.json.template` and how to ensure these are called by Nx commands.

```bash
{
  "name": "<%= name %>",
  "version": "0.0.1",
  "description": "Nx Solid App",
  "scripts": {
    "start": "vite",
    "dev": "vite",
    "build": "vite build",
    "serve": "vite preview"
  },  
}
```

DominikPieper commented:
You added the Vite nx plugin as a dependency, you need to configure the build and dev executors for the project you generate

kristianmandrup commented:
I looked at the build failure

```bash
FAIL solid packages/solid/src/generators/application/application.spec.ts
  ● Solid app schematic › Vite bundle › should add vite specific files
```

Don't see why it would fail. The files it tests for are all there as templates in `generators/application/files` including `vite.config.js` and `index.html` in the root of `files`.

kristianmandrup commented:
OK, I think I fixed the executors part. I simply copied the vite executors to solid. I also remove rollup dependencies as they are not needed by solidjs. Getting there.

DominikPieper commented:
@kristianmandrup it should work with the Vite executors, but I can take a look at that later as well and try. For now the copy is fine and that's easy to refactor or find out why later :-)

DominikPieper commented:
If you haven't seen already, there's a script in the repo called 'create-playground'. If you call it (e.g. 'yarn create-playground') it'll create a nx repo in 'tmp/nx-playground/proj' with the local versions of the plugins

kristianmandrup commented:
Awesome! Thanks @DominikPieper, I will try it out!

DominikPieper commented on packages/solid/src/generators/application/schema.d.ts:
```diff
@@ -0,0 +1,22 @@
+  directory?: string;
+  host?: string;
+  port?: number;
+  bundler?: 'vite';
```
Vite is the only option, because of that, I'd remove the bundler property. (Or do you plan more? What's supported by solidjs?

kristianmandrup replied:
Agreed. I can see however that you can also configure solid to use `esbuild`, so I left it open for the future. Will remove it for now however.

DominikPieper commented on packages/solid/src/generators/init/lib/add-dependencies.ts:
```diff
@@ -0,0 +1,11 @@
+export function updateDependencies(tree: Tree) {
+  const devDependencies = {
+    'solid-jest': '^0.2.0',
+    solid: '^1.2.5',
```
Shouldn't this be 'solid-js'? If I call the generator it's asking because the version is not found. Seems like wrong dependency

kristianmandrup replied:
Yup, correct, should be `solid-js`

DominikPieper commented on packages/solid/src/generators/application/lib/project/vite.ts:
```diff
@@ -0,0 +1,60 @@
+    build: createBuildTarget(options),
+    serve: createServeTarget(options),
+    lint: createLintTarget(options),
+    test: createTestTarget(options),
```
Remove this one. The Jest project generator will add it already and that's why the application.spec.ts fails. 'Test target already defined in architect schema' (Or something like that)

kristianmandrup replied:
Ah, cool :) Thanks!

DominikPieper replied:
Just the test part. It’ll be added by the jest generator

DominikPieper commented on packages/solid/src/generators/library/lib/add-project.ts:
```diff
@@ -0,0 +1,59 @@
+): TargetConfiguration {
+  const { libsDir } = getWorkspaceLayout(tree);
+  return {
+    executor: '@nxext/solid:build',
```
Should be @nxext/vite or you have to change the executors in package.json to the right filename. In package.json is: "builders": "./builders.json",
but you have a executors.json

kristianmandrup replied:
Ok, I changed `package.json` to:

```
  "schematics": "./generators.json",
  "executors": "./executors.json",
```

DominikPieper commented on packages/solid/src/generators/init/lib/add-dependencies.ts:
```diff
@@ -0,0 +1,11 @@
+import { addDependenciesToPackageJson, Tree } from '@nrwl/devkit';
+
+export function updateDependencies(tree: Tree) {
+  const devDependencies = {
```
The dependency 'vite-plugin-solid' is missing

kristianmandrup replied:
Nice :)

DominikPieper commented on packages/solid/src/generators/application/files/index.html.template:
```diff
@@ -0,0 +1,16 @@
+    <noscript>You need to enable JavaScript to run this app.</noscript>
+    <div id="root"></div>
+
+    <script src="/src/index.jsx" type="module"></script>
```
this should be tsx not jsx

kristianmandrup commented:
I fixed a number of errors using the `playground` as you proposed, however still having some issues.

A referencer SolidJS typescript project can be generated using `degit` with the `solidjs/templates/ts` template, such as: `npx degit solidjs/templates/ts solidts-hello-world`

Could you please have a look to see if you can spot what I'm missing. I suspect the main remaining issue is to get the `tsconfig` files right (such as the `app` extension not overriding the wrong parts but including all that it needs).

DominikPieper commented on packages/solid/src/generators/application/lib/add-linting.ts:
```diff
@@ -0,0 +1,35 @@
+    tsConfigPaths: [
+      joinPathFragments(options.projectRoot, 'tsconfig.app.json'),
+    ],
+    eslintFilePatterns: [`${options.projectRoot}/**/*.{ts,spec.ts}`],
```
```suggestion
    eslintFilePatterns: [`${options.projectRoot}/**/*.{ts,spec.ts,tsx,spec.tsx}`],
```

DominikPieper commented on packages/solid/src/generators/application/files/eslintrc.js.template:
```diff
@@ -0,0 +1,26 @@
+  "ignorePatterns": ["!**/*"],
+  "overrides": [
+    {
+      "files": ["*.ts", "*.js", "*.svelte"],
```
```suggestion
      "files": ["*.ts", "*.js", "*.tsx"],
```

DominikPieper commented on e2e/solid-e2e/tests/solid.test.ts:
```diff
@@ -0,0 +1,93 @@
+    it('should generate app into directory', async () => {
+      await runNxCommandAsync(`generate @nxext/solid:app project/ui`);
+      expect(() =>
+        checkFilesExist(`apps/project/ui/src/main.ts`)
```
```suggestion
        checkFilesExist(`apps/project/ui/src/App.tsx`)
```

DominikPieper commented on packages/solid/src/generators/library/files/eslintrc.js.template:
```diff
@@ -0,0 +1,26 @@
+  "ignorePatterns": ["!**/*"],
+  "overrides": [
+    {
+      "files": ["*.ts", "*.js", "*.svelte"],
```
```suggestion
      "files": ["*.ts", "*.js", "*.tsx"],
```

DominikPieper commented on packages/solid/src/generators/library/files/tsconfig.json.template:
```diff
@@ -0,0 +1,33 @@
+    "importsNotUsedAsValues": "error",
+    "isolatedModules": true,
+    /**
+      To have warnings/errors of the Svelte compiler at the correct position,
```
Maybe remove the hint for svelte

kristianmandrup replied:
Ah yeah, copy/paste from svelte config. I will use the same tsconfig as for application

DominikPieper commented on packages/solid/src/generators/utils/lint.ts:
```diff
@@ -0,0 +1,36 @@
+  },
+};
+
+export const createSvelteEslintJson = (projectRoot: string) => `
```
```suggestion
export const createSolidEslintJson = (projectRoot: string) => `
```

DominikPieper commented:
@kristianmandrup I added some Ideas, that should solve some of the problems. The library build problem seems like a problem with the build configuration. I'm not sure how solid libs are usually built but the vite config is missing and the vite build target searches for an index.html which isn't there.
The lint and directory generation errors were most likely configurational and with small additions I mentioned it should work. :-)

kristianmandrup commented:
Getting there :) Thanks!!!

DominikPieper commented on packages/solid/src/generators/application/application.spec.ts:
```diff
@@ -0,0 +1,40 @@
+
+  describe('Vite bundle', () => {
+    it('should add vite specific files', async () => {
+      await applicationGenerator(tree, { ...options, bundler: 'vite' });
```
```suggestion
      await applicationGenerator(tree, { ...options });
```

DominikPieper commented on packages/solid/src/generators/library/library.spec.ts:
```diff
@@ -0,0 +1,64 @@
+    await libraryGenerator(tree, options);
+    const packageJson = readJson(tree, 'package.json');
+
+    expect(packageJson.devDependencies['solid']).toBeDefined();
```
```suggestion
    expect(packageJson.devDependencies['solid-js']).toBeDefined();
```

DominikPieper commented:
@kristianmandrup after that small typos the unit tests should run fine :-)

kristianmandrup commented:
Awesome!

DominikPieper commented:
@kristianmandrup all unit tests are green, right now the buildable lib build is the last failing e2e test

kristianmandrup commented:
Awesome!

```bash
  ● solid e2e › solid lib › should be able build lib if buildable

    Command failed: yarn nx build solidlib5756888
    Could not resolve entry module (libs/solidlib5756888/index.html).
    Error during bundle: Error: Could not resolve entry module (libs/solidlib5756888/index.html).
    error Command failed with exit code 1.
```

Should be easy to fix? But should a lib even have an `index.html` file?

DominikPieper commented:
Maybe I'd add a vite config and try something like this:
```
export default defineConfig({
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.js'),
      ...
    },
    ...
})
```

kristianmandrup commented:
Yeah, I pretty figured that out myself. I added a `vite.config.js.template` to `lib/files` with 

```js
import { defineConfig } from "vite";
import solidPlugin from "vite-plugin-solid";

export default defineConfig({
  plugins: [solidPlugin()],
  build: {
    target: "esnext",
    polyfillDynamicImport: false,
    // https://vitejs.dev/guide/build.html#library-mode
    lib: {
      entry: `${options.projectRoot}/src/index.ts`,
      name: options.name,
      fileName: (format) => `${options.name}.${format}.js`,
    },
  },
});
```

DominikPieper commented:
@kristianmandrup depends on what the name represents I'd say. In some I've both cause one is the Project name and one a concrete filename

kristianmandrup commented:
But could `name` contain invalid file name characters or spacing?

DominikPieper commented:
In the end, they're just what you define. You could add the fileName generation to the options and do the calculation in the normalizeOptions function. "name" if generated through "names(options.name).fileName" so hopefully the fileName calculation with names(..) will handle illegal characters. Maybe it's no bad idea to add an explicit field for that to the Schemes to prevent refactorings around "name" break it unfortunately but that's more my taste than a rule

DominikPieper commented on packages/solid/src/generators/library/files/package.json.template:
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "<%= name %>",
+  "version": "0.0.1"
```
```suggestion
  "version": "0.0.1",
```

DominikPieper commented on packages/solid/src/generators/library/files/vite.config.js.template:
```diff
@@ -0,0 +1,16 @@
+    lib: {
+      entry: resolve(__dirname, 'src/index.js'),
+      name: options.name,
+      fileName: (format) => `${options.name}.${format}.js`,
```
```suggestion
      fileName: (format) => `<%= name %>.${format}.js`,
```
Mind typescript template string in code and generation templates with <%= %>

DominikPieper commented on packages/solid/src/generators/library/files/vite.config.js.template:
```diff
@@ -0,0 +1,16 @@
+import { defineConfig } from "vite";
+import solidPlugin from "vite-plugin-solid";
+
```
```suggestion
import { resolve } from 'path'; 

```

DominikPieper commented on packages/solid/src/generators/library/files/vite.config.js.template:
```diff
@@ -0,0 +1,16 @@
+    // https://vitejs.dev/guide/build.html#library-mode
+    lib: {
+      entry: resolve(__dirname, 'src/index.js'),
+      name: options.name,
```
```suggestion
      name: <%= name %>,
```

DominikPieper commented on packages/solid/src/generators/library/files/vite.config.js.template:
```diff
@@ -0,0 +1,16 @@
+    polyfillDynamicImport: false,
+    // https://vitejs.dev/guide/build.html#library-mode
+    lib: {
+      entry: resolve(__dirname, 'src/index.js'),
```
```suggestion
      entry: resolve(__dirname, 'src/index.ts'),
```

DominikPieper commented on e2e/solid-e2e/tests/solid.test.ts:
```diff
@@ -0,0 +1,93 @@
+      expect(result.stdout).toContain('Bundle complete');
+
+      expect(() =>
+        checkFilesExist(`dist/libs/${plugin}/bundle.js`)
```
```suggestion
        checkFilesExist(
          `dist/libs/${plugin}/bundle.es.js`,
          `dist/libs/${plugin}/bundle.umd.js`
        )
```

DominikPieper commented:
@kristianmandrup all tests green. I'd say I'll merge it and everything else you want to do in other PRs?

kristianmandrup commented:
Woa! I can hardly believe it! Thank you so much @DominikPieper for all your assistance and commit suggestions!
Super excited! Please merge it. At least we have a working base version for now that can be incrementally improved.

kristianmandrup commented:
Wuhuuuu! 🚀

DominikPieper commented:
Merged 🚀. From now I don't have to allow the GH Actions to run :-) After first merged PR it'll run them anyway.

DominikPieper commented:
@kristianmandrup I added solid to the cz-config. That means if you do "yarn commit" instead of "git commit" it'll ask some questions and let you choose solid as a project. It'll format the commit message for a format like "feat(solid): my feature".

kristianmandrup commented:
Awesome! 

I introduced `cz` (commitizen) on the Nx project that I introduced for the company I'm working for. I've also helped release `solid-bootstrap` recently, a port of Reacstrap 9 (for bootstrap 5). We use Nx and Reactstrap for that project, so we now have the essentials for potentially migrating over to SolidJS or at least trying it out in a micro frontend.

DominikPieper commented:
Later today I'll push it as the initial version to npm and tweet it. Do you have a Twitter handle to link you?

DominikPieper commented:
Oh okay :)

DominikPieper commented:
@kristianmandrup https://www.npmjs.com/package/@nxext/solid
